language-support/ts/daml-react/index.test.ts

```diff
@@ -307,236 +307,257 @@ describe('useFetchByKey', () => {
-
-    expect(closeHandler).toHaveBeenCalledTimes(0);
-    act(() => void emitter.emit('close', {code: 4000, reason: ''}));
-    expect(closeHandler).toHaveBeenCalledTimes(1);
-    expect(closeHandler).toHaveBeenLastCalledWith({code: 4000, reason: ''});
-  });
-
-  test('empty stream', () => {
-    // setup
-    const query = 'foo-query';
-    const [stream, emitter] = mockStream();
-    mockStreamQueries.mockReturnValueOnce(stream);
-    const hookResult = renderDamlHook(() => useStreamQuery(Foo, () => ({query}), [query]));
-    expect(mockStreamQueries).toHaveBeenCalledTimes(1);
-    expect(mockStreamQueries).toHaveBeenLastCalledWith(Foo, [{query}]);
-
-    // live event
-    act(() =>
-      void emitter.emit('live', [])
-    );
-
-    // no events have been emitted.
-    expect(hookResult.result.current).toEqual({contracts: [], loading:false});
-
-    // empty events
-    act(() =>
-      void emitter.emit('change', [])
-    );
-    expect(hookResult.result.current).toEqual({contracts: [], loading: false});
-  });
-
-
-  test('new events', () => {
-    // setup
-    const query = 'foo-query';
-    const [stream, emitter] = mockStream();
-    mockStreamQueries.mockReturnValueOnce(stream);
-    const hookResult = renderDamlHook(() => useStreamQuery(Foo, () => ({query}), [query]));
-    expect(mockStreamQueries).toHaveBeenCalledTimes(1);
-    expect(mockStreamQueries).toHaveBeenLastCalledWith(Foo, [{query: query}]);
-    expect(hookResult.result.current.contracts).toEqual([]);
-
-    // live event
-    act(() =>
-      void emitter.emit('live', [])
-    );
-
-    expect(hookResult.result.current.loading).toBe(false);
-
-    // one new event
-    act(() => void emitter.emit('change', ['event1']));
-    expect(hookResult.result.current.contracts).toEqual(['event1']);
-    expect(hookResult.result.current.loading).toBe(false);
-
-    // two new events replacing old one.
-    act(() => void emitter.emit('change', ['event2', 'event3']));
-    expect(hookResult.result.current).toEqual({contracts: ['event2', 'event3'], loading: false});
-  });
-
-  test('changed query triggers reload', () => {
-    // setup
-    const query1 = 'foo-query';
-    const query2 = 'bar-query';
-    const [stream, emitter] = mockStream();
-    mockStreamQueries.mockReturnValueOnce(stream);
-    const {result} = renderDamlHook(() => {
-      const [query, setQuery] = useState(query1);
-      const queryResult = useStreamQuery(Foo, () => ({query}), [query]);
-      return {queryResult, query, setQuery};
-    })
-    expect(mockStreamQueries).toHaveBeenCalledTimes(1);
-    expect(mockStreamQueries).toHaveBeenLastCalledWith(Foo, [{query: query1}]);
+function streamq<Q, LyingToTheCompiler extends {loading: boolean; contracts: readonly object[]}>(useFn: (template: Template<object>, factory: () => Q, deps: unknown[], closeHandler?: (e: StreamCloseEvent) => void) => LyingToTheCompiler, mkQuery: (s: string) => () => Q): () => void {
```

Comment:
A comment that explains why we have to lie to the compiler would be nice.

Reply:
I don't think we _have to_, but the current test code does and changing that seemed outside the scope of this PR.

The issue is that we're mocking the entire stream implementation in such a way that the return value of the (mocked) streams is never type checked. The logical type signature for the `streamq` and `streamk` functions would ensure that the values we get out of our streams are the ones expected by the `useStream*` functions, but all of the existing tests essentially return & check for `{loading: boolean; contract: payload}` instead of the proper `{loading: boolean, contract: {templateId: ..., payload: ..., ...}}`, i.e. we're completely skipping over the `CreateEvent` type.

While I do think this should be cleaned up at some point, I don't think it should be done as part of this PR. I'm also not sure there's that much to gain, so priority is unclear.

Reply:
Perfectly fine to not change this in this PR and maybe never. Can we copy your implementation into a code comment? I’m sure I’ll forget about this again otherwise.

Reply:
Added similar explanation as comments.